src/client/driver/iframe-communication/iframe-driver-adapter.js

```diff
@@ -0,0 +1,91 @@
+import DriverStatus from '../status';
+import DriverAdapter from './driver-adapter';
+import sendMessageToDriver from './send-message-to-driver';
+import MESSAGES from './messages';
+
+var Promise        = hammerhead.Promise;
+var nativeMethods  = hammerhead.nativeMethods;
+var messageSandbox = hammerhead.eventSandbox.message;
+
+const CHECK_IFRAME_EXISTENCE_INTERVAL = 1000;
+const WAIT_IFRAME_RESPONSE_DELAY      = 500;
+
+export default class IframeDriverAdapter extends DriverAdapter {
+    constructor (driverWindow, iframeWaitingTimeout, iframeSelector) {
+        super(driverWindow, iframeWaitingTimeout);
+
+        this.selector            = iframeSelector;
+        this.checkIframeInterval = null;
+    }
+
+    _ensureIframe () {
+        var iframe = domUtils.findIframeByWindow(this.window);
+
+        if (!iframe)
+            return Promise.reject(new CurrentIframeNotFoundError());
+
+        return ensureElementVisible(iframe, this.timeout, () => new CurrentIframeIsInvisibleError());
+    }
+
+    _waitForIframeRemovedOrHidden () {
+        return new Promise(resolve => {
+            this.checkIframeInterval = nativeMethods.setInterval.call(window,
+                () => ensureElement(this.selector, 0, noop, noop).catch(resolve),
+                CHECK_IFRAME_EXISTENCE_INTERVAL);
+        });
+    }
+
+    _waitForCommandResult () {
+        return new Promise(resolve => {
+            var resolved = false;
+
+            var onMessage = e => {
+                var msg = e.message;
+
+                if (msg.cmd === MESSAGES.onCommandExecuted) {
+                    messageSandbox.sendServiceMsg({ requestId: msg.requestId, cmd: MESSAGES.response }, e.source);
+                    tearDownAndResolve(msg.status);
+                }
+            };
+
+            var tearDownAndResolve = status => {
+                if (resolved)
+                    return;
+
+                resolved = true;
+                messageSandbox.off(messageSandbox.SERVICE_MSG_RECEIVED_EVENT, onMessage);
+                nativeMethods.clearInterval.call(window, this.checkIframeInterval);
+
+                resolve(status);
+            };
+
+            messageSandbox.on(messageSandbox.SERVICE_MSG_RECEIVED_EVENT, onMessage);
+
+            // NOTE: If an iframe was removed or became hidden while
+            // command executing, we consider this command is finished.
+            this
+                ._waitForIframeRemovedOrHidden()
+
+                // NOTE: wait for possible delayed iframe message
+                .then(() => delay(WAIT_IFRAME_RESPONSE_DELAY))
+                .then(() => tearDownAndResolve(new DriverStatus({ isCommandResult: true })));
+        });
+    }
+
+    executeCommand (command) {
+        // We should check iframe existence and visibility before the next command
+        // executing calling because it might be changed since the last command.
```

Comment:
```
        // We should check if the iframe is visible and exists before executing the next command,
        // because the iframe might be hidden or removed since the previous command.
```